File: backend/app/api/extensions.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends

from app.api.dependencies import get_extension_service
from app.schemas.extension import ExtensionListRead, ExtensionRead, ExtensionToggleRequest
from app.services.extension_service import ExtensionService

router = APIRouter(prefix="/extensions", tags=["extensions"])


@router.get("", response_model=ExtensionListRead)
def list_extensions(
    service: Annotated[ExtensionService, Depends(get_extension_service)],
) -> ExtensionListRead:
    return service.list_extensions()


@router.patch("/{extension_key}", response_model=ExtensionRead)
def toggle_extension(
    extension_key: str,
    payload: ExtensionToggleRequest,
    service: Annotated[ExtensionService, Depends(get_extension_service)],
) -> ExtensionRead:
    return service.set_extension_enabled(extension_key, payload)

File: frontend/src/lib/api/extensions.ts
import { requestPlatform, toPathSegment } from "../api-client";
import type {
  ExtensionListRead,
  ExtensionRead,
  ExtensionToggleRequest,
} from "../types/extension";

export function listExtensions(signal?: AbortSignal): Promise<ExtensionListRead> {
  return requestPlatform<ExtensionListRead>("/extensions", { signal });
}

export function toggleExtension(
  extensionKey: string,
  payload: ExtensionToggleRequest,
): Promise<ExtensionRead> {
  return requestPlatform<ExtensionRead>(`/extensions/${toPathSegment(extensionKey)}`, {
    body: payload,
    method: "PATCH",
  });
}

export const extensionsApi = {
  list: listExtensions,
  toggle: toggleExtension,
} as const;
